include/rx/ranges.hpp

```diff
@@ -302,6 +303,27 @@ template <class R, class = std::enable_if_t<is_input_or_sink_v<R>>>
     return input_range_iterator_end{};
 }
 
+template <class T, class Enable = void>
+struct has_advance_by : std::false_type {};
+template <class T>
+struct has_advance_by<T, std::void_t<decltype(std::declval<T&>().advance_by(size_t(0)))>> : std::true_type {};
+template <class T>
+constexpr bool has_advance_by_v = has_advance_by<T>::value;
+
+/// Advance a range by n steps, or until it reaches its end.
+///
+/// This calls R::advance_by() if available. Otherwise, it calls R::next() at most n times.
+template <class R>
+void advance_by(R& range, size_t n) {
+    if constexpr (has_advance_by_v<R>) {
+        range.advance_by(n);
+    } else {
+        for (size_t i = 0; i < n && !range.at_end(); ++i) {
```

Comment:
LIKELY(!range.at_end())

Reply:
In loops where the condition does not contain `||` I would not expect it to make a difference, since compilers should realize that the condition is inherently going to be true for all iterations except the last.